### PR TITLE
Set the JSX runtime based on runtime-specific pragmas

### LINKED ISSUE #12208: Make JSX pragmas override the configured runtime mode
## Feature Request

I'm basically requesting for what has been proposed by @nicolo-ribaudo [here](https://github.com/babel/babel/pull/11154#issuecomment-593362927) below the original PR introducing the new runtimes.

- [x] I would like to work on this feature!

**Is your feature request related to a problem?**

Currently, there is no way to override the configured runtime with JSX pragmas and I think it should be possible, they are mutually exclusive and pragmas should not only be used to configure the current runtime but they should also have the power to reconfigure which runtime is being used. This solution seems to be the most robust one from the alternatives - allowing the most flexibility.

How this has been shipped is causing problems, especially in 0-config tools like CRA. They detect which runtime should be used by inspecting the installed version of React (which is IMHO fine) and this suddenly causes errors to be thrown when old `@jsx` pragma is already used in the code. It unnecessarily bounds together upgrades of projects like CRA and Emotion (the latter is `@jsx`-based).

*Note*: this already works like this in TS 4.1 betas (at least to some extent, need to test more combinations of pragmas and options), although there are still some issues with the overall implementation there, like https://github.com/microsoft/TypeScript/issues/41146 and some warnings/errors for conflicting pragmas should be introduced. It shows though that they haven't thought about the need to restrict this which is what I'm asking for here.

**Documentation, Adoption, Migration Strategy**

Right now we have 3 situations that I'd like to make changes to:

1. `runtime: classic` (default) + `@jsxImportSource`

- **actual**: no error, no warning, the pragma is silently ignored
- **expected**: new transform being used

2. `runtime: automatic` + `@jsx` 

- **actual**: error is being thrown
- **expected**: old transform being used

3. `@jsx` + `@jsxImportSource`

- **actual**: it depends on the runtime used, so it's either using classic runtime, ignoring the new pragma or an error is being thrown
- **expected**: make it a `console.error`, I believe it should use the new transform though (so it would change the current behavior) unless you consider this being a breaking change.
- **expected in Babel 8**: throw an error

I believe that the proposed solutions for each combination **are not breaking changes**.

cc @gaearon @lunaruan

### PR DESCRIPTION
**Pull request by Andarist**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | fixes #12208
| Patch: Bug Fix?          |
| Major: Breaking Change?  |
| Minor: New Feature?      | x
| Tests Added + Pass?      | Yes
| Documentation PR Link    |
| Any Dependency Changes?  |
| License                  | MIT

The goal of this PR is to allow for **local** configuration (in the form of pragmas) to override the ones in the config. This is IMHO especially important for tools that configure Babel configs on behalf of the users and don't allow customization in this area (CRA) or it's just not immediately obvious that they are doing it (Next.js).

In my original issue, I've been claiming that this matches what TS has implemented but it turns out not to be true - or at least not fully. It is true in TS that `@jsxImportSource` overrides the classic runtime from `tsconfig.json` but it's not true that `@jsx` overrides the automatic runtime from `tsconfig.json`. I thought the latter is true as well but it has turned out that I've not been checking the emitted code for this situation but the resolution of the types and while indeed the classic runtime types have been chosen in this situation it was a bug that I have later fixed so this has never been released in the stable versions of TS.

To sum up the TS situation - currently, the automatic runtime is **always** chosen if it's configured anywhere, locally, or in the config. What should we do about this situation? Changing TS to the proposed semantic from this PR would be breaking change for them. Would it be OK for such a change to be made in TS 4.2? cc @weswigham , any thoughts about this one?

cc @wooorm

### CODE REVIEW

File: packages/babel-helper-builder-react-jsx-experimental/src/index.js [hunk 1]
```diff
@@ -138,53 +151,48 @@ You can set \`throwIfNamespace: false\` to bypass this warning.`,
...
           }
         }
 
+        if (runtimeMatch) {
+          runtime = runtimeMatch;
+        } else if (sourceMatch) {
+          runtime = "automatic";
+        } else if (jsxMatch || jsxFragMatch) {
+          runtime = "classic";
+        }
```
Comment (nicolo-ribaudo) [inline packages/babel-helper-builder-react-jsx-experimental/src/index.js:189]: I think that if `runtimeMatch === "classic" && (jsxMatch || jsxFragMatch)` or if `runtimeMatch === "automatic" && sourceMatch` we should throw an error, since you are setting two incompatible options at the same level.

**Comment by nicolo-ribaudo** [comment]
I see this PR as a bugfix (because we previously were ignoring explicit configuration pragmas), but there is a _slightly_ chance that it could break if someone relies on the old behavior. For example, this code would start using the new runtime because of the `@jsxImportSource` pragma:

```json5
// babel.config.json
{
  "plugins": [["@babel/plugin-transform-react-jsx", { pragma: "h" }]]
}
```

```js
/* @jsxImportSource preact */

<div />
```

File: packages/babel-helper-builder-react-jsx-experimental/src/index.js (hunk 1 above)
Comment (nicolo-ribaudo) [reply packages/babel-helper-builder-react-jsx-experimental/src/index.js:189]: Also, what happens if I set both `@jsxImportSource` and `@jsxPragma` in the same file?

Comment (Andarist) [reply packages/babel-helper-builder-react-jsx-experimental/src/index.js:189]: > I think that if runtimeMatch === "classic" && (jsxMatch || jsxFragMatch) or if runtimeMatch === "automatic" && sourceMatch we should throw an error, since you are setting two incompatible options at the same level.

Sure, can add that - a question is: do you plan to keep `@jsxRuntime` in Babel 8? I somewhat believe this is redundant and not needed.

**Comment by Andarist** [comment]
> I see this PR as a bugfix (because we previously were ignoring explicit configuration pragmas), but there is a slightly chance that it could break if someone relies on the old behavior. 

Yes, every bugfix is somebody's breaking change 🤪 question is if this particular "breaking change" sounds acceptable do you? I think the chances that somebody is relying on the current behavior in this regard are rather slim.

> Also, after this PR we must exactly match TypeScript's behavior. If TS won't make it's implementation of @jsxImportSource/@jsxPragma consistent, we should probably modify this PR so that it doesn't do it either.

Yes, I agree that TS & Babel should match in behavior - I hope @weswigham will get to us here to discuss this. It's a holiday season so I'm expecting that we might need to wait up to a few weeks. Don't want to rush anybody right now.

**Comment by Andarist** [comment]
@weswigham friendly 🏓

**Comment by Andarist** [comment]
@nicolo-ribaudo is there anything that can be done about this PR? could this be changed in v7? if not - could it be changed in v8?
